docs/topics/native/native-c-interop.md

```diff
@@ -646,4 +646,55 @@ fun readData(fd: Int) {
 ```
 
 Here we use service function `usePinned`, which pins an object, executes block and unpins it on normal and
-exception paths.
\ No newline at end of file
+exception paths.
+
+### Forward declarations
+
+To import forward declarations, use the `cnames` package. For example, to import a `cstructName` forward declaration
+declared in a C library with a `library.package`, use a special forward declaration package:
+`import cnames.structs.cstructName`.
+
+To transfer objects between libraries, you need to make an explicit `as` cast to and from the corresponding C
+forward declaration. Consider two cinterop libraries, one that has a forward declaration of a struct and the other
+with an actual implementation:
+
+```C
+// First C library
+#include <stdio.h>
+
+typedef struct ForwardDeclaredStruct ForwardDeclaredStruct;
+
+void consumeStruct(ForwardDeclaredStruct* s) {
+    printf("Struct consumed\n");
+}
+```
+
+```C
+// Second C library
+// Header:
+#include <stdlib.h>
+
+typedef struct {
+    int data;
+} ForwardDeclaredStruct;
```

Comment:
This is not exactly the right way to do this. For example, if you include both these libraries to the same C source file, you will get a compilation error, because the type is redefined incompatibly.

Let's get rid of `typedef` instead:
```C
// First C library
#include <stdio.h>

struct ForwardDeclaredStruct;

void consumeStruct(struct ForwardDeclaredStruct* s) {
    printf("Struct consumed\n");
}
```

```C
// Second C library
// Header:
#include <stdlib.h>

struct ForwardDeclaredStruct {
    int data;
};

// Implementation:
struct ForwardDeclaredStruct* produceStruct() {
    struct ForwardDeclaredStruct* s = malloc(sizeof(struct ForwardDeclaredStruct));
    s->data = 42;
    return s;
}
```